### PR TITLE
[fix](pipeline) Make all upstream tasks runnable if all tasks finishe…

### PR DESCRIPTION
**Pull request by Gabriel39**
…d in this pipeline

## Proposed changes

Consider 3 pipelines in this fragment (... -> join -> shuffle) :
pipeline 0 : `... -> local exchange sink`
pipeline 1 : `... -> join build (INNER JOIN)`
pipeline 2 : `local exchange source -> join probe (INNER JOIN) -> data stream sender `

Assume the JoinBuild returned 0 rows, join probe can finish directly once join build finished and do not need to wait for the `local exchange sink` finished. In this case, if pipeline 0 is blocked by a dependency for a long time, pipeline 2 should notify pipeline 0 to finish.

### CODE REVIEW

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

File: be/src/pipeline/pipeline_task.h [hunk 1]
```diff
@@ -306,11 +309,12 @@ class PipelineTask {
...
 
-    std::atomic<bool> _finalized {false};
+    std::atomic<bool> _finalized = false;
     std::mutex _dependency_lock;
 
-    std::atomic<bool> _running {false};
-    std::atomic<bool> _eos {false};
+    std::atomic<bool> _running = false;
+    std::atomic<bool> _eos = false;
+    std::atomic<bool> _wake_up_by_downstream = false;
```
Comment (yiguolei) [inline be/src/pipeline/pipeline_task.h:317]: add comment to this field
maybe rename to _downstream_eos

File: be/src/pipeline/exec/hashjoin_build_sink.cpp [hunk 2]
```diff
@@ -111,15 +111,18 @@ Status HashJoinBuildSinkLocalState::open(RuntimeState* state) {
...
     auto p = _parent->cast<HashJoinBuildSinkOperatorX>();
     Defer defer {[&]() {
         if (_should_build_hash_table && p._shared_hashtable_controller) {
             p._shared_hashtable_controller->signal_finish(p.node_id());
         }
     }};
 
-    if (!_runtime_filter_slots || _runtime_filters.empty() || state->is_cancelled()) {
-        return Status::OK();
+    if (!_runtime_filter_slots || _runtime_filters.empty() || state->is_cancelled() || !_eos) {
```
Comment (yiguolei) [inline be/src/pipeline/exec/hashjoin_build_sink.cpp:124]: 这么写可能有问题。
比如join build 和 join probe，假如说有一个优化，join probe 先读一个block，如果读取出来的为空，直接放弃join build，此时join build的close 怎么处理？

我感觉我们得依靠close的status 来处理一些事情，或者pipeline task 里保存一下down stream eos，然后把这个down stream eos 传递给pipeline task的close 函数，可能是通过exec status 来传递。

File: be/src/pipeline/exec/hashjoin_build_sink.cpp [hunk 3]
```diff
@@ -137,7 +140,7 @@ Status HashJoinBuildSinkLocalState::close(RuntimeState* state, Status exec_statu
 
     SCOPED_TIMER(_publish_runtime_filter_timer);
     RETURN_IF_ERROR(_runtime_filter_slots->publish(!_should_build_hash_table));
```
Comment (yiguolei) [inline be/src/pipeline/exec/hashjoin_build_sink.cpp:142]: 我有一个问题，为啥我们要再close 里去publish rf，而不是在sink 里，判断eos 然后publish？
感觉close 只应该是释放资源，不应该干别的啊

**Review by yiguolei — APPROVED** [review]

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**